packages/cw721/src/state.rs

```diff
@@ -0,0 +1,520 @@
+    pub explicit_content: Option<bool>,
+    pub start_trading_time: Option<Timestamp>,
+    pub royalty_info: Option<TRoyaltyInfo>,
+}
+
+impl From<CollectionInfoExtensionMsg<RoyaltyInfo>> for CollectionInfoExtension<RoyaltyInfo> {
+    fn from(ext: CollectionInfoExtensionMsg<RoyaltyInfo>) -> Self {
+        Self {
+            description: ext.description.unwrap_or_default(),
+            image: ext.image.unwrap_or_default(),
+            external_link: ext.external_link,
+            explicit_content: ext.explicit_content,
+            start_trading_time: ext.start_trading_time,
+            royalty_info: ext.royalty_info,
+        }
+    }
+}
+
+impl Validate for CollectionInfoExtension<RoyaltyInfo> {
+    /// Validates only extension, not royalty info!
+    fn validate(&self) -> Result<(), Cw721ContractError> {
+        // check description length, must not be empty and max 512 chars
+        if self.description.is_empty() {
+            return Err(Cw721ContractError::CollectionDescriptionEmpty {});
+        }
+        if self.description.len() > MAX_DESCRIPTION_LENGTH as usize {
+            return Err(Cw721ContractError::CollectionDescriptionTooLong {
+                max_length: MAX_DESCRIPTION_LENGTH,
+            });
+        }
+
+        // check images are URLs
+        Url::parse(&self.image)?;
+        if self.external_link.as_ref().is_some() {
+            Url::parse(self.external_link.as_ref().unwrap())?;
+        }
+        // validate royalty info
+        self.royalty_info
+            .as_ref()
+            .map_or(Ok(()), |r| r.validate(None))?;
+
+        Ok(())
+    }
+}
+
+impl<T> Validate for Option<T>
+where
+    T: Validate,
+{
+    fn validate(&self) -> Result<(), Cw721ContractError> {
+        match self {
+            Some(ext) => ext.validate(),
+            // no extension, nothing to validate
+            None => Ok(()),
+        }
+    }
+}
+
+impl Update<CollectionInfoExtensionMsg<RoyaltyInfo>> for CollectionInfoExtension<RoyaltyInfo> {
+    fn update(
+        &self,
+        msg: &CollectionInfoExtensionMsg<RoyaltyInfo>,
+    ) -> Result<Self, crate::error::Cw721ContractError> {
+        let mut extension = self.clone();
+        // validate royalty before updating
+        if let Some(royalty_info) = &extension.royalty_info {
+            royalty_info.validate(msg.royalty_info.clone())?;
+        }
+        extension.description = msg.description.clone().unwrap_or(self.description.clone());
+        extension.image = msg.image.clone().unwrap_or(self.image.clone());
+        extension.external_link = msg.external_link.clone().or(self.external_link.clone());
+        extension.explicit_content = msg.explicit_content.or(self.explicit_content);
+        extension.start_trading_time = msg.start_trading_time.or(self.start_trading_time);
+        extension.royalty_info = msg.royalty_info.clone().or(self.royalty_info.clone());
+        extension.validate()?;
+
+        Ok(extension)
+    }
+}
```

Comment:
... and in implementation for Update trait it does a validation. This way it is always guaranteed there is no invalid data. Custom contracts dont need to do this extra check anymore.